### ISSUE TITLE
`discouraged_optional_collection`: ignore for Apple delegate methods

### ISSUE THREAD

**Issue by djbe** [realm/SwiftLint#2298]
### Bug Report

Quite a few delegate functions from Apple provided protocols, such as `UIApplicationDelegate` trigger a warning with the `discouraged_optional_collection` rule. A user can manually add `swiftlint:disable:next` for each of these, but ideally the rule wouldn't be triggered at all for these cases. Especially as a user of [PluggableApplicationDelegate](https://github.com/fmo91/PluggableApplicationDelegate), where this is split up over multiple files.

I don't know if it's possible for swiftlint to check if a method is an implementation of a protocol method and avoid the warning in those cases, or otherwise a bunch of cases should be added to a rule for the more common Apple protocol functions.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ ./Pods/SwiftLint/swiftlint lint
...
/Users/davidjennes/Documents/Appwise/Graco/graco-ios/Application/Sources/Application Services/CoreDataConfigurationService.swift:12:49: warning: Discouraged Optional Collection Violation: Prefer empty collection over optional collection. (discouraged_optional_collection)
```

### Environment

* SwiftLint 0.26.0
* Installation method: CocoaPods
* Nested configurations: no
* Xcode 9.4.1
* Configuration:

```yml
disabled_rules:
  - line_length

opt_in_rules:
  - array_init
  - attributes
  - convenience_type
  - closure_end_indentation
  - closure_spacing
  - contains_over_first_not_nil
  - discouraged_optional_boolean
  - discouraged_optional_collection
  - empty_count
  - empty_string
  - extension_access_modifier
  - fallthrough
  - fatal_error_message
  - first_where
  - force_unwrapping
  - function_default_parameter_at_end
  - implicit_return
  - joined_default_parameter
  - literal_expression_end_indentation
  - lower_acl_than_parent
  - modifier_order
  - multiline_arguments
  - multiline_function_chains
  - multiline_parameters
  - nimble_operator
  - number_separator
  - object_literal
  - operator_usage_whitespace
  - overridden_super_call
  - override_in_extension
  - pattern_matching_keywords
  - prohibited_super_call
  - redundant_nil_coalescing
  - sorted_first_last
  - sorted_imports
  - trailing_closure
  - unavailable_function
  - unneeded_parentheses_in_closure_argument
  - vertical_parameter_alignment_on_call
  - yoda_condition

excluded:
  - Pods
  - Application/Sources/Generated

identifier_name:
  excluded:
    - vc
    - nvc

custom_rules:
  swiftgen_assets:
    name: "SwiftGen Assets"
    regex: '(?:UIImage\(named: ".+"\))|(?:#imageLiteral\(resourceName: ".+"\))|(?:UIColor\(named: ".+"\))'
    message: "Use Asset.<asset> instead"
  swiftgen_colors:
    name: "SwiftGen Colors"
    regex: '(?:#colorLiteral\(red: .+?, green: .+?, blue: .+?, alpha: .+?\))'
    message: "Use ColorName.<color> instead"
```

* Sample code:
```swift
final class CoreDataConfigurationService: NSObject, ApplicationService {
  // This triggers a violation:
  func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
    return true
  }
}
```

**Comment by marcelofabri** [comment]
> I don't know if it's possible for swiftlint to check if a method is an implementation of a protocol method and avoid the warning in those cases

Currently, that's not possible.

> a bunch of cases should be added to a rule for the more common Apple protocol functions

I think this makes sense, but should be a configuration. Would you like to try to implement that?

**Comment by djbe** [comment]
I could give it a try 😅 Mind pointing me in the right direction?

I assume the rule would have some sort of "exceptions" list, and we could provide a default to that list (that can be expanded over time). But how would this work when a user wants to add an exception in their config?
- Override the whole "exceptions" list, so it only ignores what the user has defined in their config
- Adds to the internal list of exceptions
- Or have 2 parameters: one where the user can define their exceptions, and a flag to ignore bad Apple methods
